### PR TITLE
Vapoursynth Support - Part 1

### LINKED ISSUE #173: [Feature Request] Vapoursynth Input Support
A lot of the pieces to get this working are already there, with the biggest obstacle being PySceneDetect support (although I have an idea to solve that one that I want to try out). An obvious workaround would be to just use aomenc keyframes instead, but I much prefer the speed of PySceneDetect.

There's a good chance that I'll implement this myself, as I'm working in a similar space already, but I figured I should open this issue to track support, and collect any pertinent conversation.

### PR DESCRIPTION
**Pull request by adworacz**
Supports:
  * Pyscenedetect chunking
  * aom_keyframes chunking
  * Automatic detection of input as Vapoursynth script.
  * Cleaned up fast frame counting into unified method "frame_probe_fast"

Areas for improvement:
  * Windows support
  * vspipe arg passing
  * Update README with details on Vapoursynth integration
  * Cleanup piping command generation that's sprinkled throughout the project, including moving ffmpeg pipe generation into the `ffmpeg.py` file.
  * Clean up names that include "ffmpeg" in a few places, when they aren't actually guarranteed to be ffmpeg any more (and may be vspipe instead). 

Fixes #173

### CODE REVIEW

File: Av1an/vapoursynth.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    Get frame count from vspipe.
+    :param: source: Path to input vapoursynth (vpy/py) file
+    """
+    cmd = ["vspipe", "--info", "-i", source.as_posix(), "-"]
+    r = subprocess.run(cmd, stdout=PIPE, stderr=PIPE)
+    # TODO: This regex process is overbuilt, as the output of vspipe is very simple and only has one "Frames" result.
+    matches = re.findall(r"Frames:\s*([0-9]+)\s", r.stderr.decode("utf-8") + r.stdout.decode("utf-8"))
+    return int(matches[-1])
+
+def compose_vapoursynth_pipe(source: Path, fifo: Path = None):
```
Comment (n9Mtq4) [inline Av1an/vapoursynth.py:22]:
```suggestion
def compose_vapoursynth_pipe(source: Path, fifo: Path = None):
```

File: Chunks/chunk_queue.py [hunk 2]
```diff
@@ -102,7 +102,7 @@ def create_video_queue_hybrid(args: Args, split_locations: List[int]) -> List[Ch
     """
     keyframes = get_keyframes(args.input)
 
-    end = [frame_probe_cv2(args.input)] if frame_probe_cv2(args.input) > 1 else [frame_probe(args.input)]
+    end = [frame_probe_fast(args.input, args.is_vs)]
```
Comment (n9Mtq4) [inline Chunks/chunk_queue.py:105]:
```suggestion
    end = [frame_probe_fast(args.input, args.is_vs)]
```

File: Chunks/chunk_queue.py [hunk 3]
```diff
@@ -131,6 +131,7 @@ def create_video_queue_hybrid(args: Args, split_locations: List[int]) -> List[Ch
     return chunk_queue
 
 
+#TODO: Either update the name of this func, or change the docs, or create a new func just for vapoursynth
```
Comment (n9Mtq4) [inline Chunks/chunk_queue.py:134]:
```suggestion
# TODO: Either update the name of this func, or change the docs, or create a new func just for vapoursynth
```

File: Chunks/chunk_queue.py [hunk 4]
```diff
@@ -146,24 +147,29 @@ def create_video_queue_vsffms2(args: Args, split_locations: List[int]) -> List[C
...
+        load_script = args.temp / 'split' / 'loadscript.vpy'
+        cache_file = (args.temp / 'split' / 'ffms2cache.ffindex').absolute().as_posix()
+        with open(load_script, 'w') as file:
+            file.writelines([
+                'from vapoursynth import core\n',
+                f'core.ffms2.Source("{source_file}", cachefile="{cache_file}").set_output()\n',
+            ])
+        vs_script = load_script
+
+
```
Comment (n9Mtq4) [inline Chunks/chunk_queue.py:164]:
```suggestion

```

File: Scenedetection/aom_kf.py [hunk 5]
```diff
@@ -157,7 +157,10 @@ def compose_aomsplit_first_pass_command(video_path: Path, stat_file: Path, ffmpe
     :return: ffmpeg, encode
```
Comment (n9Mtq4) [inline Scenedetection/aom_kf.py:157]:
```suggestion
    :param is_vs: is video_path a vs script
    :return: ffmpeg, encode
```

File: Av1an/utils.py [hunk 6]
```diff
@@ -58,8 +59,35 @@ def list_index_of_regex(lst: List[str], regex_str: str) -> int:
...
+        total = frame_probe(source)
+
+    return total
 
 def frame_probe_cv2(source: Path):
     video = cv2.VideoCapture(source.as_posix())
     total = int(video.get(cv2.CAP_PROP_FRAME_COUNT))
+    video.release()
     return total
+
```
Comment (n9Mtq4) [inline Av1an/utils.py:83]:
```suggestion


```

File: Av1an/utils.py [hunk 7]
```diff
@@ -58,8 +59,35 @@ def list_index_of_regex(lst: List[str], regex_str: str) -> int:
...
+    Handles vapoursynth input as well.
+    """
+    total = 0
+    if not is_vs:
+        total = frame_probe_cv2(source)
+
+    if is_vs or total < 1:
+        total = frame_probe(source)
+
+    return total
```
Comment (n9Mtq4) [inline Av1an/utils.py:77]:
```suggestion


```

File: Av1an/utils.py [hunk 8]
```diff
@@ -58,8 +59,35 @@ def list_index_of_regex(lst: List[str], regex_str: str) -> int:
             return i
     raise ValueError(f'{reg} is not in list')
 
+def frame_probe_fast(source: Path, is_vs: bool = False):
+    """
+    Consolidated function to retrieve the number of frames from the input quickly,
+    falls back on a slower (but accurate) frame count if a quick count cannot be found.
```
Comment (n9Mtq4) [inline Av1an/utils.py:65]:
```suggestion
    falls back on a slower (but accurate) frame count if a quick count cannot be found.
```

File: Av1an/ffmpeg.py [hunk 9]
```diff
@@ -8,11 +8,10 @@
 from typing import List
 from .logger import log
 
-
-def frame_probe(source: Path):
+def frame_probe_ffmpeg(source: Path):
```
Comment (n9Mtq4) [inline Av1an/ffmpeg.py:11]:
```suggestion

def frame_probe_ffmpeg(source: Path):
```

File: Scenedetection/pyscene.py [hunk 10]
```diff
@@ -1,28 +1,48 @@
...
         min_scene_len = 15
 
-    log(f'Starting PySceneDetect:\nThreshold: {threshold}, Min scene lenght: {min_scene_len}\n')
-    video_manager = VideoManager([str(video)])
+    log(f'Starting PySceneDetect:\nThreshold: {threshold}, Min scene lenght: {min_scene_len}, Is Vapoursynth input: {is_vs}\n')
+
+    if is_vs:
+        # Handling vapoursynth, so we need to create a named pipe to feed to VideoManager.
+        # TODO: Do we clean this up after pyscenedetect has run, or leave it as part of the temp dir, where it will be cleaned up later?
+        vspipe_fifo = temp / 'vspipe.y4m'
```
Comment (n9Mtq4) [inline Scenedetection/pyscene.py:29]:
```suggestion
        vspipe_fifo = temp / 'vspipe.y4m'
```

File: Av1an/split.py [hunk 11]
```diff
@@ -8,10 +8,10 @@
 from numpy import linspace
 
 from .arg_parse import Args
-from .ffmpeg import frame_probe, get_keyframes
+from .ffmpeg import get_keyframes
```
Comment (n9Mtq4) [inline Av1an/split.py:11]:
```suggestion
```

File: Av1an/vapoursynth.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
+import re
+import subprocess
+from subprocess import PIPE, STDOUT
```
Comment (n9Mtq4) [inline Av1an/vapoursynth.py:3]:
```suggestion
from subprocess import PIPE
```

**Review by n9Mtq4 — COMMENTED** [review]
Nice work!

File: Startup/setup.py [hunk 13]
```diff
@@ -117,6 +118,12 @@ def restore_term():
     args.ffmpeg_pipe = [*args.ffmpeg, *args.pix_format,
                         '-bufsize', '50000K', '-f', 'yuv4mpegpipe', '-']
 
+    # Check if input is a Vapoursynth script.
+    # TODO: Not sure how multiple input files should be handled.
+    is_vs = is_vapoursynth(args.input[0])
+    args.is_vs = is_vs
```
[the remark this answers is not in the document]
Comment (adworacz) [reply Startup/setup.py:124]: Okay, I'll give that a try. I've updated everything on my local branch per your above comments, so I'll run some tests to make sure nothing broke and then post a new rev.

I'm assuming that a squashed commit would be preferred, so I'll just do a force push to this branch HEAD.

**Review by n9Mtq4 — APPROVED** [review]

**Comment by master-of-zen** [comment]
@adworacz Great work! :medal_sports:

**Pull request merged**